File: findDuplicateSubtrees.java
import java.util.*;
// Approach 1- Inorder Traversal and HashMap
class Solution { // TC O(n) SC O(n)
    public String inorder(Node root, List<Integer> ans, HashMap<String, Integer> map){
        if (root==null) return "";
        String representation="(" + inorder(root.left, ans, map) + ")" + root.val + "(" + 
        inorder(root.right, ans, map) + ")";
        map.put(representation, map.getOrDefault(representation,0)+1);
        if (map.containsKey(representation)){
            if(map.get(representation)==2){
                ans.add(root.val);
            }
        }
        return representation;
    }
    
    public List<Integer> findDuplicateSubtrees(Node root) {
        List<Integer> ans=new ArrayList<>();
        inorder(root, ans, new HashMap<String, Integer>());
        return ans;
    }
}

class findDuplicateSubtrees{
	public static void main(String[] args) {
		BinaryTree bt= new BinaryTree();
		bt.root=new Node(1);
		bt.root.left=new Node(2);
		bt.root.right=new Node(3);
		bt.root.left.left=new Node(4);
		bt.root.right.left=new Node(2);
		bt.root.right.left.left=new Node(4);
		bt.root.right.right=new Node(4);
		Solution s= new Solution();
		System.out.println(s.findDuplicateSubtrees(bt.root));
	}
}

File: findDuplicateSubtrees.py
from collections import defaultdict
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

# Approach 1- Inorder Traversal and HashMap
class Solution:
    def findDuplicateSubtrees(self, root): # TC O(n) // SC O(n)
        def traverse(node):
            if not node:
                return ""
            representation = ("(" + traverse(node.left) + ")" + str(node.val)
                              + "(" + traverse(node.right) + ")")
            cnt[representation] += 1
            if cnt[representation] == 2:
                res.append(node.val)
            return representation
        cnt = defaultdict(int)
        res = []
        traverse(root)
        return res

root=TreeNode(1)
root.left=TreeNode(2)
root.right=TreeNode(3)
root.left.left=TreeNode(4)
root.right.left=TreeNode(2)
root.right.left.left=TreeNode(4)
root.right.right=TreeNode(4)
s=Solution()
print(s.findDuplicateSubtrees(root))

